### PR TITLE
[Eclipse] support maven project for container features

### PR DESCRIPTION
**Pull request by Eskibear**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/DockerRunHandler.java [hunk 1]
```diff
@@ -82,36 +101,54 @@ public Object onExecute(ExecutionEvent event) throws ExecutionException {
...
+            Path targetDockerfile = Paths.get(basePath, Constant.DOCKERFILE_FOLDER, Constant.DOCKERFILE_NAME);
+            ConsoleLogger.info(String.format("Validating dockerfile ... [%s]", targetDockerfile));
+            if (!targetDockerfile.toFile().exists()) {
+                throw new FileNotFoundException("Dockerfile not found.");
+            }
+            // replace placeholder if exists
+            String content = new String(Files.readAllBytes(targetDockerfile));
+            content = content.replaceAll(Constant.DOCKERFILE_ARTIFACT_PLACEHOLDER,
+                    Paths.get(basePath).toUri().relativize(Paths.get(destinationPath).toUri()).getPath());
+            Files.write(targetDockerfile, content.getBytes());
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/DockerRunHandler.java]: if file is opened, will write fail?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java [hunk 2]
```diff
@@ -53,35 +65,77 @@ public Object onExecute(ExecutionEvent event) throws ExecutionException {
...
-        DockerUtil.buildImage(dockerClient, project, project.getLocation() + Constant.DOCKER_CONTEXT_FOLDER);
-        DockerRuntime.getInstance().setLatestArtifactName(project.getName());
+        if (project == null) {
+            throw new Exception("Project is null.");
+        }
+        String basePath = project.getLocation().toString();
+        DockerClient docker = DockerRuntime.getInstance().getDockerBuilder().build();
+        // export WAR file
+
+        String destinationPath = project.getLocation() + Constant.DOCKERFILE_FOLDER + project.getName() + ".war";
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: the initial value is not used. So empty string is OK?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java [hunk 3]
```diff
@@ -53,35 +65,77 @@ public Object onExecute(ExecutionEvent event) throws ExecutionException {
...
+        Path targetDockerfile = Paths.get(basePath, Constant.DOCKERFILE_FOLDER, Constant.DOCKERFILE_NAME);
+        ConsoleLogger.info(String.format("Validating dockerfile ... [%s]", targetDockerfile));
+        if (!targetDockerfile.toFile().exists()) {
+            throw new FileNotFoundException("Dockerfile not found.");
+        }
+        // replace placeholder if exists
+        String content = new String(Files.readAllBytes(targetDockerfile));
+        content = content.replaceAll(Constant.DOCKERFILE_ARTIFACT_PLACEHOLDER,
+                Paths.get(basePath).toUri().relativize(Paths.get(destinationPath).toUri()).getPath());
+        Files.write(targetDockerfile, content.getBytes());
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: This part of code is duplicate with the one in local run. Can we abstract it?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 4]
```diff
@@ -34,50 +35,81 @@
...
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.atomic.AtomicReference;
-import org.eclipse.core.resources.IFile;
-import org.eclipse.core.resources.IFolder;
-import org.eclipse.core.resources.IProject;
-import org.eclipse.core.runtime.CoreException;
 
 public class DockerUtil {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: Can we move DockerUtil to Utils after this release?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 5]
```diff
@@ -34,50 +35,81 @@
...
-        IFile dockerfile = folder.getFile(filename);
-        if (!dockerfile.exists()) {
-            byte[] bytes = String.format(content, project.getName()).getBytes();
-            dockerfile.create(new ByteArrayInputStream(bytes), false, null);
+    }
+
+    /**
+     * runContainer.
+     */
+    public static Container runContainer(DockerClient docker, String containerId)
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null checks or add not null

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 6]
```diff
@@ -34,50 +35,81 @@
...
-     * 
-     * @param docker
-     * @param project
-     * @param imageName
-     * @return
-     * @throws DockerException
-     * @throws InterruptedException
+     * buildImage.
+     */
+    public static String buildImage(DockerClient docker, String imageNameWithTag, Path dockerDirectory,
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null checks or add notnull

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 7]
```diff
@@ -34,50 +35,81 @@
...
+            ProgressHandler progressHandler) throws DockerException, InterruptedException, IOException {
+        String imageId = docker.build(dockerDirectory, imageNameWithTag, progressHandler);
+        return imageId == null ? null : imageNameWithTag;
+    }
+
+    /**
+     * Push image to a private registry.
      */
-    public static String createContainer(DockerClient docker, IProject project, String imageName)
+    public static void pushImage(DockerClient dockerClient, String registryUrl, String registryUsername,
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null checks or add not null

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 8]
```diff
@@ -34,50 +35,81 @@
...
+            dockerClient.push(targetImageName, handler, registryAuth);
+        } else {
+            throw new DockerException("serverUrl and imageName mismatch.");
+        }
+    }
+
+    /**
+     * create container with specified ImageName:TagName.
+     */
+    public static String createContainer(DockerClient docker, String imageNameWithTag)
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null check

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 9]
```diff
@@ -87,66 +119,34 @@ public static String createContainer(DockerClient docker, IProject project, Stri
...
-            DockerRuntime.getInstance().setRunningContainerId(res.get().id());
-            return String.format("http://%s:%s", docker.getHost(),
-                    res.get().ports().stream()
-                            .filter(item -> item.privatePort().toString().equals(Constant.TOMCAT_SERVICE_PORT))
-                            .findFirst().get().publicPort());
-        } else {
-            String errorMsg = String.format(Constant.ERROR_STARTING_CONTAINER, containerId);
-            ConsoleLogger.error(errorMsg);
-            throw new Exception(errorMsg);
+    public static void stopContainer(DockerClient dockerClient, String runningContainerId)
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null check

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java [hunk 10]
```diff
@@ -87,66 +119,34 @@ public static String createContainer(DockerClient docker, IProject project, Stri
...
-            public void progress(ProgressMessage message) throws DockerException {
-                final String imageId = message.buildImageId();
-                if (imageId != null) {
-                    imageIdFromMessage.set(imageId);
-                }
-            }
-        });
-        DockerRuntime.getInstance().setLatestImageName(imageName);
-        return imageName;
+    public static DockerClient getDockerClient(String dockerHost, boolean tlsEnabled, String certPath)
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: null check?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/DockerRunHandler.java (hunk 1 above)
Comment (Eskibear) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/DockerRunHandler.java]: it won't fail.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java (hunk 4 above)
Comment (Eskibear) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/DockerUtil.java]: yes that's another task, I want to do it in another PR

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java (hunk 2 above)
Comment (Eskibear) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: yes, and I'd rather use null

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java (hunk 3 above)
Comment (Eskibear) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: low prior now

**Pull request merged**